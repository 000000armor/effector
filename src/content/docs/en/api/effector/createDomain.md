---
title: createDomain
description: createDomain is a method for creating a domain
---

Creates a [domain](/api/effector/Domain.md)

```typescript
createDomain(name?)
```

**Arguments**

1. `name`? (_string_): domain name

**Returns**

[_Domain_](/api/effector/Domain.md): New domain

### Example

```js
import { createDomain } from "effector";

const domain = createDomain(); // Unnamed domain
const httpDomain = createDomain("http"); // Named domain

const statusCodeChanged = httpDomain.createEvent();
const downloadFx = httpDomain.createEffect();
const apiDomain = httpDomain.createDomain(); // nested domain
const $data = httpDomain.createStore({ status: -1 });
```

[Try it](https://share.effector.dev/GMpjINHa)
