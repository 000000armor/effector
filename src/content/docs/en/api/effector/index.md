---
id: effector
title: API Reference
---

# List of the methods

## Unit creators

- [createEvent](./createEvent)
- [createStore](./createStore)
- [createEffect](./createEffect)
- [createDomain](./createDomain)

## Unit definitions

- [Event](./Event)
- [Effect](./Effect)
- [Store](./Store)
- [Domain](./Domain)
- [Scope](./Scope)

## Common methods

- [combine(...stores, f)](./combine)
- [attach({effect, mapParams, source?})](./attach)
- [createApi(store, api)](./createApi)
- [forward({from, to})](./forward)
- [merge([eventA, eventB])](./merge)
- [sample({clock, source, fn, target})](./sample)
- [guard({source, filter, target})](./guard)
- [split(event, cases)](./split)
- [fromObservable(observable)](./fromObservable)

## Fork API

- [fork](./fork)
- [serialize](./serialize)
- [hydrate](./hydrate)
- [allSettled](./allSettled)
- [scopeBind](./scopeBind)

## Plugins

- [effector/babel-plugin](./babel-plugin)
- [@effector/swc-plugin](https://github.com/effector/swc-plugin)

## Utilities

- [is](./is)

## Low level api

- [clearNode](./clearNode)
- [withRegion](./withRegion)
- [launch](./launch)
