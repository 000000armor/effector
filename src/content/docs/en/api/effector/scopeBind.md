---
title: scopeBind
description: scopeBind is a method to bind unit to scope
---

## Formulae {#scopeBind-formulae}

```ts
scopeBind<T>(event: Event<T>): (payload: T) => void
```

A method to bind event or effect to a [Scope](/en/api/effector/Scope) to be called later. effector supports imperative event calling inside watchers, but sometimes you should bind events to the scope explicitly, e.g. when calling events within a setTimeout/setInterval callbacks.

### Arguments {#scopeBind-arguments}

1. `event` [_Event_](/en/api/effector/Event) to be bound to the scope

## Example {#scopeBind-example}

We are going to call `changeLocation` inside `history.listen` callback so there is no way for effector to associate event with corresponding scope, and we should explicitly bind event to scope using `scopeBind`.

```js
const installHistory = app.createEvent<any>()
const changeLocation = app.createEvent<string>()

installHistory.watch(history => {

  const locationUpdate = scopeBind(changeLocation)
  history.listen(location => {
    locationUpdate(location.pathname)
  })
})
```

[See full example](https://github.com/effector/effector/blob/master/examples/react-ssr/src/app.tsx#L128)
