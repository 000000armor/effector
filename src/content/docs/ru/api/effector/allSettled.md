---
title: allSettled
lang: ru
---

::: info Нет перевода
Эта статья еще не переведена на русский язык. <br/>
Пожалуйста, используйте ссылку внизу "Edit this page", чтобы добавить перевод.<br/>
[Прочитайте оригинальную статью](/api/effector/allSettled.md) на английском.
:::
