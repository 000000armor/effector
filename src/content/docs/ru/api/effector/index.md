---
title: Справочник по API
description: Перечень основных методов API, по группам
lang: ru
---

Перечень основных методов API, по группам

## Создание юнитов

- [createEvent](./createEvent)
- [createStore](./createStore)
- [createEffect](./createEffect)
- [createDomain](./createDomain)

## Типы юнитов

- [_Event_](./Event)
- [_Effect_](./Effect)
- [_Store_](./Store)
- [_Domain_](./Domain)
- [_Scope_](./Scope)

## Основные методы библиотеки

- [combine(...stores, f)](./combine)
- [attach({effect, mapParams, source?})](./attach)
- [createApi(store, api)](./createApi)
- [forward({from, to})](./forward)
- [merge([eventA, eventB])](./merge)
- [sample({clock, source, fn, target})](./sample)
- [guard({source, filter, target})](./guard)
- [split(event, cases)](./split)
- [fromObservable(observable)](./fromObservable)

## Fork API

- [fork](./fork)
- [serialize](./serialize)
- [hydrate](./hydrate)
- [allSettled](./allSettled)
- [scopeBind](./scopeBind)

## Плагины для компилятора

- [effector/babel-plugin](./babel-plugin)
- [@effector/swc-plugin](https://github.com/effector/swc-plugin)

## Служебные функции

- [is](./is)

## Низкоуровневый API

- [clearNode](./clearNode)
- [withRegion](./withRegion)
- [launch](./launch)
