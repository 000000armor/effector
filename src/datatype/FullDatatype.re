module Cmd /*: {    type cmd;    let compute: Js.Json.t => cmd;    let run: Js.Json.t => cmd;    let emit: Js.Json.t => cmd;    let filter: Js.Json.t => cmd;    let update: Js.Json.t => cmd;    let show: cmd => string;  }*/ = {
  type cmdType =
    | ComputeType
    | EmitType
    | RunType
    | FilterType
    | UpdateType;
  let cmdT =
    fun
    | ComputeType => 11
    | EmitType => 12
    | RunType => 13
    | FilterType => 14
    | UpdateType => 15;
  let cmdUnTF =
    fun
    | 11 => ComputeType
    | 12 => EmitType
    | 13 => RunType
    | 14 => FilterType
    | 15 => UpdateType
    | _ => ComputeType;
  [@bs.deriving abstract]
  type cmd = {
    [@bs.as "type"]
    type_: int,
    data: Js.Json.t,
  };
  let cmdUnT = cmd => cmdUnTF(cmd |. type_);
  let cmdByType = (c: cmdType, data: Js.Json.t) : cmd => {
    let ins = cmd(~type_=cmdT(c), ~data);
    ins;
  };
  let compute = cmdByType(ComputeType);
  let run = cmdByType(RunType);
  let emit = cmdByType(EmitType);
  let filter = cmdByType(FilterType);
  let update = cmdByType(UpdateType);
  let show: cmd => string =
    cmd =>
      switch (cmd |. type_) {
      | 11 => "Compute"
      | 12 => "Emit(" ++ [%raw "cmd.data.fullName"] ++ ")"
      | 13 => "Run"
      | 14 => "Filter"
      | 15 => "Update"
      | _ => ""
      };
};

let cmd = {
  "compute": Cmd.compute,
  "run": Cmd.run,
  "emit": Cmd.emit,
  "filter": Cmd.filter,
  "update": Cmd.update,
  "show": Cmd.show,
  "COMPUTE": 11,
  "EMIT": 12,
  "RUN": 13,
  "FILTER": 14,
  "UPDATE": 15,
};

/*: {    type time;    type ctx('a) = {      .      "type_": int,      "data": 'a,      "time": time,    };    type computeData;    let compute:      (. array(Js.Json.t), Js.Json.t, Js.Json.t, bool, bool, bool) =>      ctx(computeData);    type emitData;    let emit: (. string, Js.Json.t) => ctx(emitData);    type filterData;    let filter: (. Js.Json.t, bool) => ctx(filterData);    type updateData;    let update: (. Js.Json.t) => ctx(updateData);    type runData;    let run: (. array(Js.Json.t), ctx(Js.Json.t)) => ctx(runData);  }*/
module Ctx = {
  type ctxType =
    | ComputeType
    | EmitType
    | RunType
    | FilterType
    | UpdateType;
  let ctxT =
    fun
    | ComputeType => 21
    | EmitType => 22
    | RunType => 23
    | FilterType => 24
    | UpdateType => 25;
  type time;
  /* type date; abstract type for a document object */
  [@bs.scope "Date"] [@bs.val] external now : unit => time = "now";
  [@bs.deriving abstract]
  type ctx('a) = {
    [@bs.as "type"]
    type_: int,
    data: 'a,
    time,
  };
  let getType = (ctx: ctx('a)) => ctx |. type_;
  [@bs.deriving abstract]
  type computeData = {
    args: array(Js.Json.t),
    result: Js.Json.t,
    error: Js.Json.t,
    isError: bool,
    isNone: bool,
    isChanged: bool,
  };
  let compute =
    (.
      args: array(Js.Json.t),
      result: Js.Json.t,
      error: Js.Json.t,
      isError: bool,
      isNone: bool,
      isChanged: bool,
    ) => {
      let data =
        computeData(~args, ~result, ~error, ~isError, ~isNone, ~isChanged);
      ctx(~type_=ctxT(ComputeType), ~time=now(), ~data);
    };
  [@bs.deriving abstract]
  type runData = {
    args: array(Js.Json.t),
    parentContext: ctx(Js.Json.t) /* compute | emit | filter | update */
  };
  let run =
    (. args: array(Js.Json.t), parentContext: ctx(Js.Json.t)) => {
      let data = runData(~args, ~parentContext);
      ctx(~type_=ctxT(RunType), ~time=now(), ~data);
    };
  [@bs.deriving abstract]
  type emitData = {
    eventName: string,
    payload: Js.Json.t,
  };
  let emit =
    (. eventName: string, payload: Js.Json.t) => {
      let data = emitData(~payload, ~eventName);
      ctx(~type_=ctxT(EmitType), ~time=now(), ~data);
    };
  [@bs.deriving abstract]
  type filterData = {
    value: Js.Json.t,
    isChanged: bool,
  };
  let filter =
    (. value: Js.Json.t, isChanged: bool) => {
      let data = filterData(~value, ~isChanged);
      ctx(~type_=ctxT(FilterType), ~time=now(), ~data);
    };
  [@bs.deriving abstract]
  type updateData = {value: Js.Json.t};
  let update =
    (. value: Js.Json.t) => {
      let data = updateData(~value);
      ctx(~type_=ctxT(UpdateType), ~time=now(), ~data);
    };
  type ctxg =
    | ComputeG(ctx(computeData))
    | EmitG(ctx(emitData))
    | RunG(ctx(runData))
    | FilterG(ctx(filterData))
    | UpdateG(ctx(updateData));
};

let getCtxType = Ctx.getType;

let ctx = {
  "compute": Ctx.compute,
  "run": Ctx.run,
  "emit": Ctx.emit,
  "filter": Ctx.filter,
  "update": Ctx.update,
  "COMPUTE": 21,
  "EMIT": 22,
  "RUN": 23,
  "FILTER": 24,
  "UPDATE": 25,
};

module Step /*: {    type step('a);    /* [@bs...] external step : (~type_: int, ~data: 'a) => step('a) = ""; */    /* external type_ : step('a) => int = "";       external data : step('a) => 'a = ""; */    /* module Label:       {         type stepType = SingleType | MultiType | SeqType;         let stepT: stepType => int;         let single: int;         let multi: int;         let seq: int;       }; */    type set('a) = {. add: (. 'a) => set('a)};    let setAdd = (s: set('a), a: 'a) => s |. add(a);    /* [@bs...] external createSet: array('a) => set('a) = "Set"; */    let single: (~data: Cmd.cmd) => step(Cmd.cmd);    let multi: unit => step(set('a));    let seq: (~data: array(step('a))) => step(array(step('a)));    let show: step('a) => string;  }*/ = {
  [@bs.deriving abstract]
  type step('a) = {
    [@bs.as "type"]
    type_: int,
    data: 'a,
  };
  module Label = {
    type stepType =
      | [@bs.as 31] SingleType
      | [@bs.as 32] MultiType
      | [@bs.as 33] SeqType;
    let stepT =
      fun
      | SingleType => 31
      | MultiType => 32
      | SeqType => 33;
    let single = stepT(SingleType);
    let multi = stepT(MultiType);
    let seq = stepT(SeqType);
  };
  type set('a) = {. add: (. 'a) => set('a)};
  let setAdd = (s: set('a), v: 'a) => {
    s;
    v;
    %raw
    "s.add(v)";
    ();
  };
  [@bs.new] external createSet : array('a) => set('a) = "Set";
  let single = (~data: Cmd.cmd) => step(~data, ~type_=Label.single);
  let multi = () => step(~data=createSet([||]), ~type_=Label.multi);
  let seq = (~data: array(step(_))) => step(~data, ~type_=Label.seq);
  [@bs.module "./showstep.js"]
  external show' : (step('a), Cmd.cmd => string) => string = "show";
  let show = a => show'(a, Cmd.show);
};

let step = {
  "single": (~data: Cmd.cmd) => Step.single(~data),
  "multi": () => (Step.multi(): Step.step(Step.set(int))),
  "seq": (~data: array(Step.step(int))) => Step.seq(~data),
  "show": (a: Step.step(int)) => Step.show(a),
  "SINGLE": 31,
  "MULTI": 32,
  "SEQ": 33,
};