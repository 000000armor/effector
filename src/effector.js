//@flow

import type {Store} from 'redux'
import {Stream, from} from 'most'
import {async as subject, type Subject} from 'most-subject'

import type {ID, Tag, Domain, Effect, Event} from './index.h'
import {counter, toTag} from './index.h'

export function createDomain<State>(
  store: Store<State>,
  domainName: string = '~'
): Domain<State> {
  const redux: any = store
  const state$: Stream<State> = from(redux)
    .map(() => store.getState())
    .multicast()
  return DomainConstructor(
    domainName,
    redux.dispatch,
    store.getState,
    state$,
  )
}

export function domain<State>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
): Domain<State> {
  return DomainConstructor(
    domainName,
    dispatch,
    getState,
    state$,
  )
}


const nextPayloadID = counter()
const nextEventID = counter()

function DomainConstructor<State>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events = new Map
): Domain<State> {
  return {
    effect<Params, Done, Fail>(
      name: string
    ): Effect<Params, Done, Fail, State> {
      return EffectConstructor(
        domainName,
        dispatch,
        getState,
        state$,
        events,
        name
      )
    },
    domain(name: string) {
      return DomainConstructor(
        [domainName, name].join('/'),
        dispatch,
        getState,
        state$,
        events,
      )
    },
    event<Payload>(
      name: string
    ): Event<Payload, State> {
      return EventConstructor(
        domainName,
        dispatch,
        getState,
        state$,
        events,
        name
      )
    }
  }
}

function EffectConstructor<State, Params, Done, Fail>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events,
  name: string
): Effect<Params, Done, Fail, State> {
  const handlers = new Set
  const eventID = nextEventID()
  const nextSeq = counter()
  const action$: Subject<Params> = subject()
  const getType = () => toTag(domainName, name)
  handlers.add((payload, state) => action$.next(payload))
  // handlers.add((payload, state) => state$.next(state))
  let using: (params: Params) => Promise<Done> = never
  function use(thunk: (params: Params) => Promise<Done>) {
    using = thunk
  }
  const create = (payload: Params) => ({
    type: getType(),
    payload,
    meta: {
      index: nextPayloadID(),
      eventID,
      seq: nextSeq(),
    },
  })
  const effect = (params: Params) => {
    let run = () => {
      run = () => {}
      using(params).then(
        result => {
          const data = {params, result}
          dispatch(done(data))
          resolve(data)
        },
        error => {
          const data = {params, error}
          dispatch(fail(data))
          reject(data)
        },
      )
    }
    const runner = (dispatchHook) => {
      if (canDispatch) {
        const dispatcher = dispatchHook
          ? dispatchHook
          : dispatch
        canDispatch = false
        dispatcher(result)
        for (const handler of handlers) {
          handler(params, getState())
        }
        action$.next(params)
        run()
      }
    }
    let resolve = (val) => {}
    let reject = (val) => {}
    const doneP: Promise<{params: Params, result: Done}> = new Promise(rs => resolve = rs)
    const failP: Promise<{params: Params, error: Fail}> = new Promise(rs => reject = rs)
    const toSend = Promise.resolve(params)
    const result = create(params)
    let canDispatch = true
    return {
      ...result,
      raw() {
        return result
      },
      send(dispatchHook) {
        runner(dispatchHook)
        return toSend
      },
      done() {
        runner()
        return doneP
      },
      fail() {
        runner()
        return failP
      },
      promise() {
        runner()
        return Promise.race([
          doneP,
          failP.then(data => Promise.reject(data)),
        ])
      }
    }
  }
  const done: Event<{params: Params, result: Done}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'done'].join(' ')
  )
  const fail: Event<{params: Params, error: Fail}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'fail'].join(' ')
  )
  function watch<R>(fn: (params: Params, state: State) => R) {
    handlers.add((params, state) => {
      const result = fn(params, state)
      safeDispatch(result, dispatch)
    })
  }
  function epic<R>(
    handler: (
      data$: Stream<Params>,
      state$: Stream<State>
    ) => Stream<R>
  ): Stream<R> {
    const result = handler(action$, state$)
      .multicast()
    result.observe(data => safeDispatch(data, dispatch))
    return result
  }
  effect.use = use
  effect.getType = getType
  effect.done = done
  effect.fail = fail
  effect.watch = watch
  effect.epic = epic
  return effect
}

function safeDispatch(data, dispatch) {
  if (
    typeof data === 'object'
    && data != null
    && typeof data.type === 'string'
  ) {
    dispatch(data)
  }
}

function EventConstructor<State, Payload>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events,
  name: string
): Event<Payload, State> {
  const eventID = nextEventID()
  const nextSeq = counter()
  const getType = () => toTag(domainName, name)
  const handlers = new Set
  function watch<R>(
    fn: (params: Payload, state: State) => R
  ) {
    handlers.add(fn)
  }
  const action$: Subject<Payload> = subject()
  handlers.add((payload, state) => action$.next(payload))
  // handlers.add((payload, state) => state$.next(state))
  function epic<R>(
    handler: (
      data$: Stream<Payload>,
      state$: Stream<State>
    ) => Stream<R>
  ): Stream<R> {
    const result = handler(action$, state$)
      .multicast()
    result.observe(data => safeDispatch(data, dispatch))
    return result
  }
  const create = (payload: Payload) => ({
    type: getType(),
    payload,
    meta: {
      index: nextPayloadID(),
      eventID,
      seq: nextSeq(),
    },
  })
  const eventInstance = (payload: Payload) => {
    const toSend = Promise.resolve(payload)
    const result = create(payload)
    let canDispatch = true
    return {
      ...result,
      raw() {
        return result
      },
      send(dispatchHook) {
        if (canDispatch) {
          const dispatcher = dispatchHook
            ? dispatchHook
            : dispatch
          canDispatch = false
          dispatcher(result)
          for (const handler of handlers) {
            handler(payload, getState())
          }
          action$.next(payload)
        }
        return toSend
      },
    }
  }
  eventInstance.getType = getType
  //$off
  eventInstance.toString = getType
  eventInstance.watch = watch
  eventInstance.epic = epic

  events.set(getType(), eventInstance)
  return eventInstance
}


function never(props: any): Promise<any> {
  console.warn(`


  Running an effect without implementation

`, props)
  return new Promise(() => {})
}


