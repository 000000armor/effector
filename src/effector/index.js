//@flow

/*
 * People assume that time is a strict progression of cause to effect,
 * but actually from a non-linear, non-subjective viewpoint -
 * it's more like a big ball of wibbly wobbly time-y wimey stuff
 */

export {combine, combine as createStoreObject} from './combine'
export {sample} from './sample'

export type {Domain} from './domain'
export {createDomain} from './domain'

export type {Event} from './event'
export {createEvent, merge, forward, fromObservable} from './event'

export type {Effect} from './effect'
export {createEffect} from './effect'

export type {Store} from './store'
export {
  createStore,
  setStoreName,
  extract,
  createApi,
  restore,
  restoreEvent,
  restoreEffect,
  restoreObject,
  withProps,
} from './store'

export {
  Kind,
  clearNode,
  createGraph as createNode,
  step,
  traverse,
} from './stdlib'
export {launch} from './kernel'
export type {kind} from './stdlib'

export {
  invariant,
  warning,
  isUnit,
  isStore,
  isEvent,
  isEffect,
  isDomain,
  is,
} from './validate'
export {version} from 'effector/flags'

//eslint-disable-next-line
export * as blocks from './blocks'
