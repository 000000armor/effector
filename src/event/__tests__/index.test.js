//@flow

import {from} from 'most'
import {createEvent, type Event} from '..'

import {getSpyCalls, spy} from 'effector/fixtures'
// import {show} from 'effector/datatype/step/show'

import {step} from 'effector/datatype/FullDatatype.bs'
const show = step.show

describe('symbol-observable support', () => {
  test('most.from(event) //stream of events', () => {
    expect(() => {
      from(createEvent(''))
    }).not.toThrow()
    const ev1 = createEvent('ev1')
    const ev2 = createEvent('ev2')
    const ev1$ = from(ev1)
    ev1$.observe(spy)
    ev1(0)
    ev1(1)
    ev1(2)
    ev2('should ignore')
    expect(spy).toHaveBeenCalledTimes(3)

    expect(spy.mock.calls).toEqual([[0], [1], [2]])
  })
})

test('event.watch(fn)', () => {
  const click = createEvent('click')
  click.watch(spy)
  click()
  click(1)
  click(2)
  expect(spy).toHaveBeenCalledTimes(3)
  expect(getSpyCalls()).toEqual([
    [undefined, 'click'],
    [1, 'click'],
    [2, 'click'],
  ])
})

test('event.filter should infer type', () => {
  const num: Event<number | '-1'> = createEvent('number')

  const evenNum = num.filter(n => {
    if (n !== '-1') return n
  })

  evenNum.watch(e => spy(e))

  num(0)
  num('-1')
  num(2)
  num('-1')
  num(4)
  ;(evenNum: Event<number>) //Should not fail

  expect(getSpyCalls()).toEqual([[0], [2], [4]])
})

test('event.filter should drop undefined values', () => {
  const num: Event<number> = createEvent('number')
  const evenNum = num.filter(n => {
    if (n % 2 === 0) return n * 2
  })

  evenNum.watch(e => spy(e))

  num(0)
  num(1)
  num(2)
  num(3)
  num(4)

  expect(getSpyCalls()).toEqual([[0], [4], [8]])

  expect(show(num.graphite.seq)).toMatchSnapshot('num event graph')
  expect(show(evenNum.graphite.seq)).toMatchSnapshot('evenNum event graph')
})
