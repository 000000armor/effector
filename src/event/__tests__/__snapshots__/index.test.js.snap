// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event.filter should drop undefined values: evenNum event graph 1`] = `
"Seq [
  1. emit {{
     \\"subtype\\": \\"event\\",
     \\"fullName\\": \\"number →? *\\"
  }},
  2. [
     *  run {{}}
  ]
]"
`;

exports[`event.filter should drop undefined values: num event graph 1`] = `
"Seq [
  1. emit {{
     \\"subtype\\": \\"event\\",
     \\"fullName\\": \\"number\\"
  }},
  2. [
     *  Seq [
        1. compute {{}},
        2. filter {{}},
        3. Seq [
           1. emit {{
              \\"subtype\\": \\"event\\",
              \\"fullName\\": \\"number →? *\\"
           }},
           2. [
              *  run {{}}
           ]
        ]
     ]
  ]
]"
`;
